src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs

```diff
@@ -0,0 +1,262 @@
+                        this.State = new LocalState(
+                            BitVector.Create(this.State.Assigned.Capacity));
+
+                        // Assign all parameters of all parent functions
+                        EnterAllParameters();
+
+                        VisitLambdaOrLocalFunction(localFunc,
+                            recordAssigns: false);
+
+                        this.State = savedState;
+                    }
+                }
+            } while (HasDirtyLocalFunctions(_readVars.Values));
+
+            _recordingReads = recordingState;
+        }
+
+        private void EnterAllParameters()
+        {
+            foreach (var info in variableBySlot)
+            {
+                if (info.Symbol?.Kind == SymbolKind.Parameter)
+                {
+                    EnterParameter((ParameterSymbol)info.Symbol);
+                }
+            }
+        }
+
+        private void RecordReadInLocalFunction(int slot)
+        {
+            var localFunc = GetNearestLocalFunctionOpt(currentMethodOrLambda);
+
+            Debug.Assert(localFunc != null);
+
+            LocalFuncInfo readInfo;
+            if (_readVars.TryGetValue(localFunc, out readInfo))
+            {
+                if (!readInfo.UsedVars.Contains(slot))
+                {
+                    readInfo.IsDirty = true;
+                    readInfo.UsedVars.Add(slot);
+                }
+            }
+            else
+            {
+                readInfo = new LocalFuncInfo();
+                readInfo.IsDirty = true;
+                readInfo.UsedVars.Add(slot);
+                _readVars.Add(localFunc, readInfo);
+            }
+        }
+
+        /// <summary>
+        /// At the local function's use site, checks that all variables read
+        /// are assigned and assigns all variables that are definitely assigned
+        /// to be definitely assigned.
+        /// </summary>
+        protected virtual void ReplayReadsAndWrites(
+            LocalFunctionSymbol localFunc,
+            CSharpSyntaxNode syntax,
+            bool writes)
+        {
+            _usedLocalFunctions.Add(localFunc);
+
+            // First process the reads
+            LocalFuncInfo info;
+            if (_readVars.TryGetValue(localFunc, out info))
+            {
+                foreach (var usedVarSlot in info.UsedVars)
+                {
+                    var symbol = variableBySlot[usedVarSlot].Symbol;
+                    CheckAssigned(symbol, syntax, usedVarSlot);
+                }
+            }
+
+            // Now the writes
+            if (writes &&
+                _localFuncAssignmentResults?.AssignedVars
+                    .TryGetValue(localFunc, out info) == true)
+            {
+                foreach (var usedVarSlot in info.UsedVars)
+                {
+                    SetSlotAssigned(usedVarSlot);
+                }
+            }
+        }
```

Comment:
Consider extracting helper methods `ReplayReads(_readVars, localFunc)` and `ReplayWrites(_localFuncAssignmentResults.AssignedVars, localFunc)`, and use `ReplayWrites` in the derived class implementation of `ReplayReadsAndWrites`.